src/DynamoDBMessageRepository/DynamoDBMessageRepository.php

```diff
@@ -0,0 +1,140 @@
+<?php
+
+namespace EventSauce\MessageRepository\DynamoDBMessageRepository;
+
+use Aws\DynamoDb\DynamoDbClient;
+use Aws\DynamoDb\Marshaler;
+use Aws\ResultPaginator;
+use EventSauce\EventSourcing\AggregateRootId;
+use EventSauce\EventSourcing\Message;
+use EventSauce\EventSourcing\MessageRepository;
+use EventSauce\EventSourcing\Serialization\MessageSerializer;
+use EventSauce\EventSourcing\Header;
+use EventSauce\EventSourcing\UnableToPersistMessages;
+use EventSauce\EventSourcing\UnableToRetrieveMessages;
+use Ramsey\Uuid\Uuid;
+use Generator;
+
+use Throwable;
+
+class DynamoDBMessageRepository implements MessageRepository
+{
+    protected DynamoDbClient $client;
+    protected MessageSerializer $serializer;
+    protected string $tableName;
+
+    public function __construct(DynamoDbClient $client, MessageSerializer $serializer, string $tableName)
+    {
+        $this->client = $client;
+        $this->serializer = $serializer;
+        $this->tableName = $tableName;
+    }
+
+    public function persist(Message ...$messages): void
+    {
+        if (count($messages) === 0) {
+            return;
+        }
+
+        $items = [];
+        $marshaler = new Marshaler();
+
+        foreach ($messages as $message) {
+            $payload = $this->serializer->serializeMessage($message);
+            $event = $payload['headers'][Header::EVENT_ID] = $payload['headers'][Header::EVENT_ID] ?? Uuid::uuid4()->toString();
+            $aggregateRootId = $payload['headers'][Header::AGGREGATE_ROOT_ID] ?? null;
+            $eventType = $payload['headers'][Header::EVENT_TYPE] ?? null;
+            $aggregateRootVersion = $payload['headers'][Header::AGGREGATE_ROOT_VERSION] ?? 0;
+            $timeOfRecording = $payload['headers'][Header::TIME_OF_RECORDING];
+
+            $item = [
+                'event' => $marshaler->marshalValue($event),
+                'aggregateRootId' => $marshaler->marshalValue($aggregateRootId),
+                'eventType' => $marshaler->marshalValue($eventType),
+                'aggregateRootVersion' => $marshaler->marshalValue($aggregateRootVersion),
+                'timeOfRecording' => $marshaler->marshalValue($timeOfRecording),
+                'payload' => $marshaler->marshalItem(['payload' => $payload])['payload']
+            ];
+
+            $items[] = [
+                'PutRequest' => [
+                    'Item' => $item
+                ]
+            ];
+        }
+
+        $batchRequest = ['RequestItems' => [ $this->tableName => $items ]];
+
+        try {
+            $this->client->batchWriteItem($batchRequest);
+        } catch (Throwable $exception) {
+            throw UnableToPersistMessages::dueTo('', $exception);
```

Comment:
I guess a consumer would/should expect that no events have been stored / your event store is still in the same state, but that's not necessarily true. You could exceed your provisioned DynamoDB capacity halfway through the batch, leaving you with a partial update.